lib/widgets/share.dart

```diff
@@ -0,0 +1,83 @@
+import 'dart:async';
+
+import 'package:flutter/material.dart';
+import 'package:flutter/scheduler.dart';
+
+import '../model/narrow.dart';
+import 'compose_box.dart';
+import 'icons.dart';
+import 'message_list.dart';
+import 'page.dart';
+import 'recent_dm_conversations.dart';
+import 'subscription_list.dart';
+
+class SharePage extends StatelessWidget {
+  const SharePage({
+    super.key,
+    required this.filesToUpload,
+    required this.sharedText,
+  });
+
+  final Iterable<FileToUpload>? filesToUpload;
+  final String? sharedText;
+
+  static AccountRoute<void> buildRoute({
+    required int accountId,
+    required Iterable<FileToUpload>? filesToUpload,
+    required String? sharedText,
+  }) {
+    return MaterialAccountWidgetRoute(
+      accountId: accountId,
+      page: SharePage(
+        filesToUpload: filesToUpload,
+        sharedText: sharedText));
+  }
+
+  void _handleNarrowSelect(BuildContext context, Narrow narrow) {
+    final composeBoxStateKey = GlobalKey<ComposeBoxState>();
+
+    unawaited(Navigator.pushReplacement(context,
+      MessageListPage.buildRoute(
+        context: context,
+        narrow: narrow,
+        composeBoxStateKey: composeBoxStateKey)));
+
+    SchedulerBinding.instance.addPostFrameCallback((_) async {
+      final composeBoxState = composeBoxStateKey.currentState;
+
+      // TODO show an error, probably the message list doesn't have a ComposeBox.
+      if (composeBoxState == null) {
+        return; // TODO(log)
+      }
+
+      if (filesToUpload != null) {
+        await composeBoxState.uploadFiles(filesToUpload!);
+      } else if (sharedText != null) {
+        composeBoxState.insertText(sharedText!);
+      }
```

Comment:
Does this need the `else`? It seems like this could work just fine with both `if` blocks acting independently.

Probably then put the text before the uploads.

I'm imagining that there might be sharing sources that include both some text and some images or other files. I'm not sure if the platform APIs support that; but it's definitely a natural thing for people to want to do, so perhaps they might. (After all, even when the thing shared from the other app is purely an image, people often/usually want to add to add some text before sending it, in Zulip as well as in many other apps.)